### DISCUSSION TITLE
Material Properties as a function of temperature

### DISCUSSION THREAD

**Discussion by alikhan-zada** [idaholab/moose#17820] [Q&A General] [answered]
Hello everyone, 
I'm new to MOOSE and I am currently trying to use the Heat Conduction Module. Please bear with my novice question even if they might be very obvious

1. I want to be able to use the thermal conductivity and specific heat as a function of temperature, something of this sort:

thermal conductivity = 11.82 + 0.0106 T
specific heat = 330.9 + 0.563 T - 0.0004015 T^2 + 0.00000009465 T^3

I have tried doing this using **thermal_conductivity_temperature_function** in the materials block, but Functions can only take in only (x,y,z,t) as variables. 

[Functions]
  [./k_func]
    type = ParsedFunction
    vars = temp
    value = '11.82 + 0.0106 * temp'
  [../]
  [./c_func]
    type = ParsedFunction
    vars = temp
    value = '330.9 + 0.563 * temp - 0.0004015 * temp * temp + 0.00000009465 *temp * temp * temp'
  [../]
[] # Functions

[Materials]
  [./heat]
    type = ADHeatConductionMaterial
    block = 0
    temp = temp
    thermal_conductivity_temperature_function = k_func
    specific_heat_temperature_function = c_func
  [../]
[]
_Which is obviously wrong. Can someone point me in the right direction, please :)_

2. My other question is, how can I link the FunctionPathEllipsoidHeatSource and grain_growth_2D_graintracker together?
I know there's an implementation of the grain_growth_2D_graintracker where the mobility is a function of temperature. This can be done using multiapps, but which multiapp would be the most suitable for this application? A question similar to this ([here](https://github.com/idaholab/moose/discussions/17708)) has already been asked in the discussions but the answers are not very helpful. 

3. Lastly, what are the units of the different constants used in  FunctionPathEllipsoidHeatSource?

Kind regards,
Ali

Edit: **_I was able to link up the two apps, so need to worry about question 2_**  😅️
Edit 2: for question 3, units of material properties in the function_ellipsoid_heat_source are as follows: 
Specific heat = J/(Kg . K)
Thermal Conductivity = W/(mm . K)
Density = Kg/mm^3

Please correct me if I am wrong. 😅

**Comment by aeslaughter** [comment, marked as answer]
I would create a new Material that defines the conductivity and specific heat with your desired function coded in the `Material` object.

We do this in Step 5 and 6 of the YouTube tutorial on the front page of mooseframework.org. Click the 'SHOW MORE" and you will be able to go to the section.

In the tutorial we do a linear interpolation based on tabulated data to compute the temperature dependent properties: https://github.com/idaholab/moose/blob/next/tutorials/darcy_thermo_mech/step06_coupled_darcy_heat_conduction/src/materials/PackedColumn.C

**Reply by alikhan-zada** [reply]
Thank you Andrew,
I'll take a look and let you know if I have any more questions.

Cheers,
Ali

**Reply by alikhan-zada** [reply]
Hey @aeslaughter, so what if I want to deposit a layer of material after each time the heat source passed from on top of the material?
Something like this https://ars.els-cdn.com/content/image/1-s2.0-S0924013618300980-mmc4.gif

**Reply by dewenyushu** [reply]
We do have some developments for this purpose. I would suggest you to look at: https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html

The idea is to divide your physical domain into "active" and "inactive" subdomains, define your physical kernels in the "active subdomain" (with 'block = #'), and constantly moving elements from  "inactive subdomain" -> "active subdomain". (Set the `kernel_coverage_check = false` in the `Problem` block if no kernel is defined in the inactive subdomain)

To realize layer-by-layer material addition, I would try to set the `coupled_var` to be the z-position of the element. The threshold value to be the height of the heat source (I guess), below which the elements will be added. If you would like to apply BC at the top of the product as it is growing, then set the `moving _boundary_name` to be the sideset of the product top surface.

@hugary1995 please feel free to add/correct me here since you made some great changes to this code :)

**Reply by alikhan-zada** [reply]
Thank you @dewenyushu for your answer.
As you can see in the gif, there is also a 2-D sectioned area, is it possible to deploy a MultiApp system just in that 2-D section and transfer the value of the heat source?

What I am guessing is that both the Master and the sub app use the TransientMultiApp system and since the scales are not that different use the MultiAppCopyTransfer because I will probably use the same mesh distribution. But I might need to use a different transfer function if I decide to use adaptivity, what would be the best choice in that case?

**Reply by dewenyushu** [reply]
Since I am not an expert on multiapp and transfers, @jiangwen84 would you mind answering this question?
